internal/k8s/object.go

```diff
@@ -0,0 +1,105 @@
+package k8s
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type ObjectClient struct {
+	k8sClient client.Client
+	log       *logr.Logger
+}
+
+func NewObjectClient(k8sClient client.Client, log *logr.Logger) (*ObjectClient, error) {
+	if k8sClient == nil {
+		return nil, errors.New("k8sClient cannot be nil")
+	}
+	if log == nil {
+		return nil, errors.New("log cannot be nil")
+	}
+	return &ObjectClient{
+		k8sClient: k8sClient,
+		log:       log,
+	}, nil
+}
+
+// EnsureFinalizer adds `finalizer` to obj exactly once.
+//   - It NO-OPs if the finalizer is already there.
+//   - It uses a strategic-merge Patch so you don’t overwrite concurrent changes.
+//   - obj must be a pointer that already contains the latest copy from the API server.
+func (s *ObjectClient) EnsureFinalizer(ctx context.Context, key client.ObjectKey, finalizer string) error {
+	var obj client.Object
+	err := s.k8sClient.Get(ctx, key, obj)
+	if err != nil {
+		return err
+	}
+	// if it already contains the finalizer, there's nothing to do
+	if controllerutil.ContainsFinalizer(obj, finalizer) {
+		return nil
+	}
+
+	controllerutil.AddFinalizer(obj, finalizer)
+
+	base, err := s.deepCopyObject(obj)
+	if err != nil {
+		return err
+	}
+
+	s.log.
+		WithValues("finalizer", finalizer).
+		WithValues("object", fmt.Sprintf("%s/%s", obj.GetNamespace(), obj.GetName())).
+		Info("creating finalizer")
+	if err := s.k8sClient.Patch(ctx, obj, client.MergeFrom(base)); err != nil {
```

Comment:
Any way to share this with the existing Merge/Apply in `kubectl_apply.go`? Ok to resolve if not too easy.

Reply:
yeah I'll have a look into this

Reply:
okay I had a look at this, and my gut feeling is that not having this depending on the genericReconciler is a good thing, it's pretty self contained as it stands and I think we can do work to break down the genericReconciler interface into being multiple leaner interfaces. What do you think?